### PR TITLE
fix: prevent overwriting template files via `<FileTree>`

### PR DESCRIPTION
**Pull request by AriPerkkio**
Show an error if user tries to create file that already exists on file system, but is not visible on FileTree. Such files are for example non-lesson template files.

### CODE REVIEW

File: packages/runtime/src/store/tutorial-runner.ts [hunk 1]
```diff
@@ -189,6 +189,62 @@ export class TutorialRunner {
...
+        async (signal) => {
+          await previousLoadPromise;
+
+          const webcontainer = await this._webcontainer;
+
+          if (signal.aborted) {
+            reject(new Error('Task was aborted'));
+          }
+
+          signal.throwIfAborted();
```
Comment (AriPerkkio) [inline packages/runtime/src/store/tutorial-runner.ts:206]: @Nemikolh, extra attention here please 😄 
Is this correct?

File: packages/types/src/entities/index.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@ export type * from './nav.js';
 export type FileDescriptor = { path: string; type: 'file' | 'folder' };
 export type Files = Record<string, string | Uint8Array>;
 
+export type FilesystemError = 'FILE_EXISTS' | 'FOLDER_EXISTS';
```
Comment (AriPerkkio) [inline packages/types/src/entities/index.ts:9]: I thought about using real `Error` here, like `class FilesystemError extends Error {}` but maybe we can use something as simple as string literals here? Any thoughts?

File: packages/runtime/src/store/tutorial-runner.ts (hunk 1 above)
Comment (Nemikolh) [reply packages/runtime/src/store/tutorial-runner.ts:206]: It is! :raised_hands: 

Thinking more about this, given what this method does, I think that we should just remove everything related to cancellation here. It's mostly there to speed up when moving between lessons quickly.

File: packages/runtime/src/store/tutorial-runner.ts [hunk 3]
```diff
@@ -189,6 +189,62 @@ export class TutorialRunner {
...
+          } catch {
+            resolve(false);
+          }
+        },
+        { ignoreCancel: true },
+      );
+    });
+  }
+
+  async folderExists(folderPath: string) {
```
Comment (Nemikolh) [inline packages/runtime/src/store/tutorial-runner.ts:220]: Oh maybe we could create a helper function that both `folderExists` and `fileExists` call? Wdyt?

File: packages/types/src/entities/index.ts (hunk 2 above)
Comment (Nemikolh) [reply packages/types/src/entities/index.ts:9]: I think this is fine, we could also use an enum but a string literal is fine by me :+1:

File: packages/react/src/core/ContextMenu.tsx [hunk 4]
```diff
@@ -72,14 +75,22 @@ export function ContextMenu({
...
+      try {
+        await onFileChange?.({
           value,
           type: state === 'add_file' ? 'file' : 'folder',
           method: 'add',
         });
-      } else {
-        return setState('add_failed');
+      } catch (error: any) {
+        const message: FilesystemError | (string & {}) | undefined = error?.message;
```
Comment (Nemikolh) [inline packages/react/src/core/ContextMenu.tsx:89]: Oh first time seeing this kind of types. What benefit does `FilesystemError | (string & {}) | undefined` has?

**Review by Nemikolh — COMMENTED** [review]
Changes looks good!

Seeing the screenshot, I wonder if we should add more padding on the y-axis? What do you think?

File: packages/react/src/core/ContextMenu.tsx (hunk 4 above)
Comment (AriPerkkio) [reply packages/react/src/core/ContextMenu.tsx:89]: It's for IDE auto completion: https://www.typescriptlang.org/play/?#code/C4TwDgpgBA0gdgewO5wCrggZygXigIgDMBLAJ02HygB8DMIBjBOAE3wG4AoT0SKAQQA2g5EMEBlYKWJwA5rliIU6SNloVpcmlACurCCTgQWXJnApREASTjAIw4vXMQAXAOGjhkzfLz4O3LzQYp6C-KwACqQIAG7ELBD8OsAIAMIIALZgghB2CvDIaBhqUBoy8gBkUADeAL7aegmGxqbMFgAWAIaYNnYOTvRuIUhi4SxRsfGJyWmZ2bnQfgGcAPQrUACysdCECAw62ABGuUgQEHAEVJ2sUFLEsrIQpFCdM1BMWTl5SMTA7VCpDYAEQA1JgwJ0GNB1qlUAAlAAyYIhUM4QA

Here we expect that any kind of errors can come up, but want to make sure we don't typo types of `FilesystemError`.

Comment (Nemikolh) [reply packages/react/src/core/ContextMenu.tsx:89]: Wooo this is super coool! :star_struck: TIL.

File: packages/runtime/src/store/tutorial-runner.ts [hunk 5]
```diff
@@ -189,6 +189,62 @@ export class TutorialRunner {
     );
   }
 
+  async fileExists(filepath: string) {
+    const previousLoadPromise = this._currentLoadTask?.promise;
```
[the remark this answers is not in the document]
Comment (AriPerkkio) [reply packages/runtime/src/store/tutorial-runner.ts:193]: Good point! I'll add that. 👍

**Comment by AriPerkkio** [comment]
> Seeing the screenshot, I wonder if we should add more padding on the y-axis?

Sure! Now it looks like this:

**Review by Nemikolh — APPROVED** [review]
Good job on this! :tada:

**Pull request merged**